finetrainers/trainer.py

```diff
@@ -0,0 +1,542 @@
+                    noisy_latents = (1.0 - sigmas) * latent_conditions["latents"] + sigmas * noise
+                    
+                    latent_conditions.update({"noisy_latents": noisy_latents})
+                    other_conditions.update({"timesteps": timesteps})
+
+                    # These weighting schemes use a uniform timestep sampling and instead post-weight the loss
+                    weights = compute_loss_weighting_for_sd3(weighting_scheme=self.args.flow_weighting_scheme, sigmas=sigmas)
+                    pred = self.model_config["forward_pass"](transformer=self.transformer, **latent_conditions, **other_conditions)
+                    target = noise - latent_conditions["latents"]
+                    
+                    loss = weights.float() * (pred["latents"].float() - target.float()).pow(2)
+                    # Average loss across channel dimension
+                    loss = loss.mean(list(range(1, loss.ndim)))
+                    # Average loss across batch dimension
+                    loss = loss.mean()
+
+                    accelerator.backward(loss)
+                    
+                    self.optimizer.step()
+                    self.lr_scheduler.step()
+                    self.optimizer.zero_grad()
+                
+                # Checks if the accelerator has performed an optimization step behind the scenes
+                if accelerator.sync_gradients:
+                    progress_bar.update(1)
+                    global_step += 1
+
+                    # Checkpointing
+                    if accelerator.is_main_process or accelerator.distributed_type == DistributedType.DEEPSPEED:
+                        logger.info(f"Checkpointing at step {global_step}")
+                        if global_step % self.args.checkpointing_steps == 0:
+                            # _before_ saving state, check if this save would set us over the `checkpointing_limit`
+                            if self.args.checkpointing_limit is not None:
+                                checkpoints = os.listdir(self.args.output_dir)
+                                checkpoints = [d for d in checkpoints if d.startswith("checkpoint")]
+                                checkpoints = sorted(checkpoints, key=lambda x: int(x.split("-")[1]))
+
+                                # before we save the new checkpoint, we need to have at _most_ `checkpoints_total_limit - 1` checkpoints
+                                if len(checkpoints) >= self.args.checkpointing_limit:
+                                    num_to_remove = len(checkpoints) - self.args.checkpointing_limit + 1
+                                    removing_checkpoints = checkpoints[0:num_to_remove]
+
+                                    logger.info(
+                                        f"{len(checkpoints)} checkpoints already exist, removing {len(removing_checkpoints)} checkpoints"
```

Comment:
This could be factored out into a separate method.

Reply:
Cleaned it up a bit, but yeah will refactor the entire thing in a follow-up as it's not the most important thing